feat(devops): Dockerhub Added latest, edge and feature branch for dockerhub

daehli opened the pull request:
Added the latest tag for the docker hub via Github action.

I will update the documentation to talk about the latest tags.

## Example

Every branch will be push to the dockerhub registry.  If the branch is named `feature/00-modified-documentation`

The tag in the dockerhub will be `feature-00-modified-documentation`. It can be useful to test a branch for a PR really quickly. The only thing to do is to pull from the register

```
docker pull botpress/server:feature-00-modified-documentation
```

For example, I created 2 tags in one build to show the behavior of the code. I'm building a list of tags and I'm pushing it to the dockerhub registery.

![image](https://user-images.githubusercontent.com/6981314/127048887-a9a41ac5-9a50-41a2-bb76-27cdb1020efd.png)


## Example
The dockerHub tags was created by the github action
![image](https://user-images.githubusercontent.com/6981314/126994281-3633fb65-44ff-4adf-aa9a-a26994eb90aa.png)


![image](https://user-images.githubusercontent.com/6981314/127012057-db8e0562-9171-4231-a77e-8c23a8452744.png)


## UNIT-Test 

The documentation for the push event $GITHUB_REF is a little bit obscure. I tried a different set-up and I was not able to get the tags. I came with a different approach. I getting the tags from the SHA hash.

```
[Docker/Build and push Docker image to GitHub Packages] ⭐  Run Prepare
[Docker/Build and push Docker image to GitHub Packages]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/botpress/github/botpress/botpress/workflow/prep] user=
[Module Builder Docker Image/Test Module Builder Image] ⭐  Run Build
[Module Builder Docker Image/Test Module Builder Image]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/botpress/github/botpress/botpress/workflow/1] user=
| VALID TAG
[Docker/Build and push Docker image to GitHub Packages]   ⚙  ::set-output:: version=100.0.2 latest v100_0_2
[Docker/Build and push Docker image to GitHub Packages]   ⚙  ::set-output:: tags=botpress/server:100.0.2,botpress/server:latest,botpress/server:v100_0_2
[Docker/Build and push Docker image to GitHub Packages]   ⚙  ::set-output:: created=2021-07-27T19:56:07Z
```

The tags push to the dockerhub will be 

```
botpress/server:100.0.2
botpress/server:latest
botpress/server:v100_0_2
```


## Further action

- Create a Clean-up function to remove older tags.
- Reduce the size of the image
- Migrate from `master` to `main` branch name

daehli committed:
Added latest,edge and feature branch for dockerhub

daehli committed:
# This is a combination of 19 commits.
# This is the 1st commit message:

Added Push to the dockerhub registry

# This is the commit message #2:

Modified context for dockerhub build

# This is the commit message #3:

Adding Array expensiend in the Prepare job.

Created an array of tags to push to the dockerhub registery

# This is the commit message #4:

Added an array test to github action

# This is the commit message #5:

Repo name was not formatted correctly

# This is the commit message #6:

Array of element

# This is the commit message #7:

From the documentation of the buildx docker.

The array of tags need to be in the array string format

# This is the commit message #8:

Transform tags as string

# This is the commit message #9:

Removed the option parameters

The default option is the value I use

# This is the commit message #10:

Modified join to the array

# This is the commit message #11:

Transform the Array into a StringArray

# This is the commit message #12:

Remove test for multiple tags

# This is the commit message #13:

Modified the dockerhub readme

# This is the commit message #14:

Addedd more section to the dockerhub readme

# This is the commit message #15:

Modified the Documentation

# This is the commit message #16:

Updated the Documentation for docker

# This is the commit message #17:

Added backward compatibility for the release branch

# This is the commit message #18:

Test if the tags is triggered

# This is the commit message #19:

Modified the PUSH and BUILD function

charlescatta commented:
This looks great, just wanted to know the behavior in a specific case:

What happens if I create a branch named `v12_24_0`, will the docker image tag be:
`feature-v12_24_0` or `v12_24_0` ?

If it is `v12_24_0` we have to be careful when creating the branch for a release as `botpress/server:v12_24_0` will point to the branch that's preparing the version instead of the future master commit tagged with `v12_24_0`!

daehli commented:
If you create a branch without tags named `v12_24_0`. The tag in dockerhub will be `botpress/server:v12_24_0`

The code

```
elif [[ $GITHUB_REF == refs/heads/* ]]; then
  VERSION+=($(echo ${GITHUB_REF#refs/heads/} | sed -r 's#/+#-#g')) # transform feature/00-feature-branch to feature-00-feature-branch
fi
```

will transform a branch named `feature/00-feature` to `botpress/server:feature-00-feature`.

A branch Named `hotfix/00-hotfix` will be tagged as `botpress/server:hotfix-00-hotfix`. 

The `/` character is not a valid character for the docker tags. I transform the branch name to be a valid tag.

## Commit with Tags 

If a commit is pushed with a tag Ex .`v12.24.0`. The tag in dockerhub will be like this; `botpress/server:12.24.0`

```
          if [[ $GITHUB_REF == refs/tags/* ]]; then
            VERSION+=($(echo ${GITHUB_REF#refs/tags/} | sed -r 's/v//'')) # transform v12.23.0 to 12.23.0
            VERSION+=("latest")
```

I use the same naming convention as MySQL, grafana and elasticsearch. I change the way we name think for out release (v12_23_0) to have a clear cut between the old deployment methods and the new ones. We can easily know when the new Github action is trigger.

What do you think about this ? 

**Elasticsearch**
![image](https://user-images.githubusercontent.com/6981314/127176651-bf3f3fac-9bd9-491e-9005-133055f52f49.png)

**Mysql**
![image](https://user-images.githubusercontent.com/6981314/127176323-d535d1d8-440c-4d2c-a77f-e97cf65ea9d6.png)

**Grafana**

![image](https://user-images.githubusercontent.com/6981314/127176508-599ca1f5-0492-447d-bb07-085f4284ccdf.png)

daehli commented on .github/workflows/docker.yml:
```diff
@@ -1,10 +1,38 @@
+          VERSION=()
+          if [[ $GITHUB_REF == refs/tags/* ]]; then
+            VERSION+=($(echo ${GITHUB_REF#refs/tags/} | sed -r 's/v//'')) # transform v12.23.0 to 12.23.0
+            VERSION+=("latest")
```
added backward compatibility

charlescatta commented:
Looks good !

daehli committed:
Added Push to the dockerhub registry

Modified context for dockerhub build

Adding Array expensiend in the Prepare job.

Created an array of tags to push to the dockerhub registery

Added an array test to github action

Repo name was not formatted correctly

Array of element

From the documentation of the buildx docker.

The array of tags need to be in the array string format

Transform tags as string

Removed the option parameters

The default option is the value I use

Modified join to the array

Transform the Array into a StringArray

Remove test for multiple tags

Modified the dockerhub readme

Addedd more section to the dockerhub readme

Modified the Documentation

Updated the Documentation for docker

Added backward compatibility for the release branch

Test if the tags is triggered

Modified the PUSH and BUILD function

chore(dev): reorganize repository (#5153)

Added Push to the dockerhub registry

Modified context for dockerhub build

Adding Array expensiend in the Prepare job.

Created an array of tags to push to the dockerhub registery

Added an array test to github action

Repo name was not formatted correctly

Array of element

From the documentation of the buildx docker.

The array of tags need to be in the array string format

Transform tags as string

Removed the option parameters

The default option is the value I use

Modified join to the array

Transform the Array into a StringArray

Remove test for multiple tags

Modified the dockerhub readme

Addedd more section to the dockerhub readme

Modified the Documentation

Updated the Documentation for docker

Added backward compatibility for the release branch

Test if the tags is triggered

Modified the PUSH and BUILD function

chore(dev): reorganize repository (#5153)

laurentlp reviewed (approved):
LGTM! Good job!

daehli committed:
Merge branch 'master' into feature/00-dockerhub-tag
